File: src/ParallelFuzzyMatch/ParallelPatterns.FuzzyMatch/Module1/TaskComposition.cs
using System;
using System.Threading.Tasks;

namespace ParallelPatterns.TaskComposition
{
    public static class TaskComposition
    {
        // TODO (1)
        // implement missing code
        public static Task<TOut> Then<TIn, TOut>(
            this Task<TIn> task,
            Func<TIn, TOut> next)
        {
            var tcs = new TaskCompletionSource<TOut>();

            // Missing code

            return tcs.Task;
        }

        // TODO (1)
        // implement missing code
        public static Task<TOut> Then<TIn, TOut>(
            this Task<TIn> task,
            Func<TIn, Task<TOut>> next)
        {
            var tcs = new TaskCompletionSource<TOut>();

            // Missing code

            return tcs.Task;
        }


        public static Task<TOut> Select<TIn, TOut>(
            this Task<TIn> task,
            Func<TIn, TOut> projection)
        {
            var r = new TaskCompletionSource<TOut>();

            // Missing code

            return r.Task;
        }
    }
}

File: src/ParallelFuzzyMatch/ParallelPatterns.FuzzyMatch.FSharp/Module1/TaskComposition.fs
namespace ParallelPatterns.Fsharp
open System
open System.Threading.Tasks
open System.Runtime.CompilerServices

[<Extension>]
module TaskCompositionEx =
    [<Extension>]
    type TaskEx =
        [<Extension>]
        static member Then (input : Task<'T>, 
                            binder :Func<'T, Task<'U>>) =
            let tcs = new TaskCompletionSource<'U>()
            
            // missing code

            tcs.Task
            
        // TODO (1)
        // implement missing code
        [<Extension>]
        static member  Then (input : Task<'T>, 
                             binder :Func<'T, Task<'U>>, 
                             projection:Func<'T, 'U, 'R>) =
            let tcs = new TaskCompletionSource<'U>()
            
            // missing code

            tcs.Task

        // TODO (1)
        // implement missing code
        [<Extension>]
        static member Then (input : Task<'T>, 
                            binder :Func<'T, 'U>) =
            let tcs = new TaskCompletionSource<'U>()
            
            // missing code

            tcs.Task


        [<Extension>]
        static member Select (task : Task<'T>, selector :Func<'T, 'U>) : Task<'U> =
            let tcs = new TaskCompletionSource<'U>()
            
            // missing code

            tcs.Task
